adding backup targets

TAlonglong opened the pull request:
<!-- Describe what your PR does, and why -->
<!-- For works in progress choose "Create draft pull request" from the drop-down green button. -->

 - [x] Closes #169  <!-- remove if there is no corresponding issue, which should only be the case for minor changes -->
 - [x] Tests added  <!-- for all bug fixes or enhancements -->
 - [x] Fully documented <!-- remove if this change should not be visible to users, e.g., if it is an internal clean-up, or if this is part of a larger project that will be documented later -->
 - [x] Add your name to `AUTHORS.md` if not there already

pnuu reviewed (approved):
LGTM. Just couple of small thoughts that can be ignored.

TAlonglong commented on trollmoves/movers.py:
```diff
@@ -329,7 +330,11 @@ def open_connection(self):
+                destination = backup_targets.pop(0)
+                LOGGER.info("Changing destination to backup target: %s", destination)
+                retries = 3
+        raise IOError(f"Failed to ssh connect after 3 attempts.")
```
Yeah, I was thinking something like that. You see I have the `f` I added and the forgot about it. I will a suggestion

mraspaud commented on trollmoves/tests/test_ssh_server.py:
```diff
@@ -175,6 +175,32 @@ def test_scp_open_connection_socket_timeout_exception(self, mock_sshclient_conne
+        try:
+            with self.assertLogs(logger, level=logging.INFO) as lc, self.assertRaises(IOError):
+                scp_mover.open_connection()
+            self.assertIn(("SSH connection timed out:"), lc.output[0])
+            self.assertIn(("Changing destination to backup target: backup_host1"), lc.output[3])
+            self.assertIn(("Changing destination to backup target: backup_host2"), lc.output[7])
```
Now that we are using pytest, the preferred way to do this is
`assert "SSH connection timed out:" in lc.output[0]`

Also caplog could be used, but no rush for that.

mraspaud reviewed (approved):
LGTM, just one styling comment.

pnuu commented:
One thing I forgot to ask for: example configuration. Which made me realise that the original primary `destination` is configured in the Client side, but the backups need to be in the Server config, right? Should the backup servers be sent in the request message by the Client? That way the Server process could be used also for other transfers and it wouldn't fallback to push the data to wrong backup hosts.

TAlonglong commented:
Ah, crap. You are right.

TAlonglong commented:
The implementation I came up with does not work at all.

TAlonglong commented:
So, I think I will revert all of this work and start over. Maybe I'll start a new PR to add the `start()` to the publisher to be compatible with posttroll 1.10.0

Then I need a sanity check. I think I will implement the backup target if I hit here: https://github.com/pytroll/trollmoves/blob/main/trollmoves/client.py#L671

So the idea is then to, if backup targets/destinations config is provided, restart the request.

Any ideas?

pnuu commented:
I would keep the structure you have here, but instead of putting the backup destinations to Server config, put them in Client config and add them in the request message maybe in [this](https://github.com/pytroll/trollmoves/blob/main/trollmoves/client.py#L277) else block. Then in Server check the message for `backup_destinations` list in the message data if the retries go to zero.

TAlonglong commented:
Ah, yes. That looks doable. 

So, where is this tested then. Then I can try from there

TAlonglong commented:
Updated to have the backup targets config in the client config.
This is then sent to the server, which change the destination if the primary fails.
It's tricky to propagate this info from the message to the actual mover, so I added a extra function parameter for this.
One issue I see is that the final message contains the primary target, I have not found a way to replace that information in the message.

Please let me know id you see anything odd

pnuu commented:
> One issue I see is that the final message contains the primary target, I have not found a way to replace that information in the message.

This is a tricky one. The message should maybe be updated [here](https://github.com/pytroll/trollmoves/blob/main/trollmoves/server.py#LL129C15-L129C15), but that would require rather large changes so that the actual destination were passed down from the individual `Mover` to `movers.move_it()` to `server.RequestManager._move_file()` to `server.RequestManager._move_files()` and eventually to `movers.RequestManager.push()`. And then there's the complication that `_move_files()` already returns either an error message or `None`.

pnuu commented on trollmoves/tests/test_client.py:
```diff
@@ -167,6 +167,17 @@
+topic = /1b/hrit-segment/0deg
+publish_port = 0
+nameservers = ns1
+backup_targets=backup_host1 backup_host2
```
These hosts need to have the same protocol as the original destination.

TAlonglong replied:
Yeah, and that reminds me of adding an example explaining this.

pnuu commented on trollmoves/tests/test_client.py:
```diff
@@ -917,6 +934,37 @@ def test_request_push_single_call(send_ack, send_request, clean_ongoing_transfer
+    from trollmoves.client import request_push
+    from tempfile import gettempdir
+
+    MSG_FILE_BACKUP_TARGETS = MSG_FILE2
```
```suggestion
    msg_file_backup_targets = MSG_FILE2
```
This should be lower case, and adjusted accordingly also later in the code.

TAlonglong replied:
Hm, ok. I was just following the syntax used in the rest of the similar setup.

pnuu reviewed (approved):
Just couple of small comments inline, otherwise LGTM.

The comment on the config made me think that *later* we could move the retry logic to `movers.move_it()` so that all of the movers would have the possibility. Then we could also have backup destinations with different transport schemes; the primary could be `scp://` and backups `scp://` and `ftp://` for example. But that's for another PR.

TAlonglong commented:
I'm quite happy with this. Please let me know what you think

pnuu commented on trollmoves/movers.py:
```diff
@@ -71,7 +71,12 @@ def move_it(pathname, destination, attrs=None, hook=None, rel_path=None):
 
     try:
-        mover(pathname, new_dest, attrs=attrs).copy()
+        m = mover(pathname, new_dest, attrs=attrs, backup_targets=backup_targets)
+        m.copy()
+        last_dest = m.destination
```
Ah! Nice catch!

pnuu replied:
... although this only changes the log message, not the message published by the Client :see_no_evil:

pnuu replied:
One possibility:
* in `movers.move_it()` return `m.destination`
* in `server.RequestManager._move_file()` replace `message.data['destination']` with the return value and return the modified message if the transfer was succesful
  * in the case of error do what is done already
* in `server.RequestManager._move_files()` after the `self._move_file()` call check `message.type is 'err'` instead of `error_message is not None`
* in `server.RequestManager.push()` again check the `message.type`, run `clean_url` if necessary

TAlonglong replied:
Ah, you are right. Crap.

pnuu commented on trollmoves/movers.py:
```diff
@@ -329,7 +343,13 @@ def open_connection(self):
-        raise IOError("Failed to ssh connect after 3 attempts")
+            if retries == 0 and backup_targets:
+                backup_target = backup_targets.pop(0)
+                self.destination = self.destination._replace(netloc=f"{self.destination.username}@{backup_target}")
```
I was going to say using private methods might not be a good idea, but apparently this is *public* API. From https://docs.python.org/3/library/collections.html#collections.namedtuple

> To prevent conflicts with field names, the method and attribute names start with an underscore.

TAlonglong replied:
Yeah I also figured out that.

TAlonglong commented:
A step further. What do you think?

pnuu commented on trollmoves/server.py:
```diff
@@ -129,28 +129,28 @@ def pong(self, message):
         """Reply to push request."""
         new_msg = self._move_files(message)
-        if new_msg is None:
+        if new_msg.type is not 'err':
```
The default return value is `None`, so this will fail with `AttributeError`. Maybe this is safer?
```suggestion
        if new_msg and new_msg.type is not 'err':
```

pnuu commented on trollmoves/server.py:
```diff
@@ -129,28 +129,28 @@ def pong(self, message):
         new_msg = self._move_files(message)
-        if new_msg is None:
+        if new_msg.type is not 'err':
+            _destination = clean_url(new_msg.data['destination'])
```
Isn't the destination already cleaned in `_move_file()`? Or actually, maybe safer to do the cleaning here if the `err` message also has some credentials or such that needs to be cleaned.

pnuu reviewed (commented):
Couple more things. And a live test would be great :grin:

pnuu commented:
No need to demonstrate :joy: Just run a test that this works in real life.

pnuu reviewed (approved):
LGTM.

Does @mraspaud want to have a look?

TAlonglong commented:
Then I see something that might be a feature. First time a backup target is used, the active connection is tagged with the primary target. Therefore the next time, if the previous connection is still alive, this is used even if the actual connection is to the backup target. SO that way if this occur it looks like the primary target is used but the actual cache connection is to the bakcup target.

changing line https://github.com/pytroll/trollmoves/blob/main/trollmoves/movers.py#L136 to
`self.active_connections[(self.destination.hostname, port, username)] = connection, timer` will solve this, but I'm not sure how the behaviour should be

TAlonglong commented:
Not sure if this is ready to merge as I see the approval from @mraspaud is old/changes after the approval.

lahtinep commented:
I'm merging this, we need a new release with the `.start()` method to make PyPI installation work with the latest Posttroll.
